posts/14-stable-structures.html.pm

```diff
@@ -0,0 +1,671 @@
+type TxIndex = StableBTreeMap<Principal, StableVector<Transaction>>;
+}
+◊source-code["good"]{
+◊em{// ◊b{GOOD}: use a composite key to avoid nesting.}
+◊em{// Use a ◊a[#:href "#range-scan-example"]{◊b{range scan}} to find all transactions of a principal.}
+type TxIndex = StableBTreeMap<(Principal, TxId), Transaction>;
+}
+}
+
+◊p{
+  Let us examine the available data structures in detail.
+}
+
+◊subsection-title["stable-cell"]{Stable cell}
+◊p{
+  A ◊a[#:href "https://docs.rs/ic-stable-structures/latest/ic_stable_structures/cell/struct.Cell.html"]{Cell} represents a single value stored in stable memory in serialized form.
+  Cell's contents in stable memory updates every time you change the underlying value.
+}
+◊figure{
+◊marginnote["mn-log-interface"]{
+  The core interface of the ◊code{Cell} stable structure.
+}
+◊source-code["rust"]{
+impl<T: ◊code-ref["#storable-trait"]{Storable}, M: ◊code-ref["#memory"]{Memory}> struct ◊b{Cell}<T, M> {
+    ◊em{/// Returns the current cell value.}
+    ◊em{/// Complexity: O(1).}
+    pub fn ◊b{get}(&self, idx: usize) -> Option<Vec<u8>>;
+
+    ◊em{/// Updates the cell value.}
+    ◊em{/// Complexity: O(value size).}
+    pub fn ◊b{set}(&mut self, value: T) -> Result<T, ValueError>;
+}
+}
+}
+◊figure[#:class "grayscale-diagram"]{
+  ◊marginnote["mn-cell-figure"]{
+    A ◊code{Cell} persists a single value in stable memory and caches it on the heap.
+    The serialized value and the value on the heap are always in sync.
+  }
+  ◊(embed-svg "images/14-cell.svg")
+}
+◊p{
+  The primary use case for cells is storing canister configuration:
+}
+◊ul[#:class "arrows"]{
+  ◊li{The ICRC-1 Ledger Archive canister persists its initialization arguments in a ◊a[#:href "https://github.com/dfinity/ic/blob/9cdb1e62bcd199f28ae0005ed3f762487a1454df/rs/rosetta-api/icrc1/archive/src/main.rs#L49"]{cell}.}
+  ◊li{The Internet Identity Archive canister holds its init arguments in a ◊a[#:href "https://github.com/dfinity/internet-identity/blob/b66fe925fb0a337b09aaaa5beaf1a60994b19f14/src/archive/src/main.rs#L85"]{cell}.}
+}
+
+◊subsection-title["stable-vec"]{Stable vector}
+◊p{
+  A ◊code{Vec} is a growable mutable array similar to ◊code-ref["https://doc.rust-lang.org/std/vec/struct.Vec.html"]{std::vec::Vec}.
+  A stable vector stores its items by value, so it must know how much space it needs to allocate for each item, hence the ◊code-ref["#bounded-storable-trait"]{BoundedStorable} bound on the item type.
+}
+◊p{
+  Stable vector takes advantage of the ◊code-ref["#is-fixed-size-attribute"]{T::IS_FIXED_SIZE} attribute of the item type.
+  If the value size is not fixed, the vector allocates a few extra bytes to store the actual entry size and the ◊code-ref["#max-size-attribute"]{T::MAX_SIZE} bytes required for each value.
+  If all the values are the same size, the vector implementation uses ◊code-ref["#max-size-attribute"]{T::MAX_SIZE} for the item slot, saving up to 4 bytes per entry.
```

Comment:
```suggestion
  If all the values are the same size, the vector implementation uses ◊code-ref["#max-size-attribute"]{T::MAX_SIZE} for the item slot, saving 4 bytes per entry.
```

Reply:
The vector uses 1 byte for the length if max_size is at most u8::MAX and 2 bytes if max_size is at most u16::MAX.